backend/migrations/dataall_migrations/migrationmanager.py

```diff
@@ -0,0 +1,124 @@
+from typing import Type, Union
+
+import logging
+
+logger = logging.getLogger()
+logger.setLevel(os.environ.get('LOG_LEVEL', 'INFO'))
+
+
+class MigrationManager:
+    def __init__(self, current_revision='0', initial_migration=InitMigration):
+        self.initial_migration = initial_migration
+        self.previous_migrations: Deque[Union[Type[MigrationBase], None]] = deque()
+        self.current_migration = initial_migration
+
+        while self.current_migration.revision_id() != current_revision:
+            self.previous_migrations.append(self.current_migration)
+            self.current_migration = self.current_migration.next_migration()
+            if not self.current_migration:
+                raise Exception(f'Revision {current_revision} is not found.')
+
+    def save_upgraded(self, executed_ups: Deque[Union[Type[MigrationBase], None]]):
+        while executed_ups:
+            self.previous_migrations.append(executed_ups.popleft())
+
+    def save_downgrades(self, executed_downs: Deque[Union[Type[MigrationBase], None]]):
+        while executed_downs:
+            try:
+                self.previous_migrations.remove(executed_downs.pop())
+            except Exception as e:
+                ...
+
+    def check_downgrade_id(self, target_revision_id):
+        for pm in self.previous_migrations:
+            if pm.revision_id() == target_revision_id:
+                return True
+        if target_revision_id == self.current_migration.revision_id():
+            return False
+
+        raise Exception(f'Failed to find {target_revision_id} in migration history.')
+
+    def check_upgrade_id(self, target_revision_id):
+        if target_revision_id is None:
+            if (
+                self.current_migration.next_migration() is None
+                or self.current_migration.revision_id() == target_revision_id
+            ):
+                return False
+            return True
+
+        revision = self.current_migration.next_migration()
+        while revision is not None:
+            if revision.revision_id() == target_revision_id:
+                return True
+            revision = revision.next_migration()
+
+        Exception(f'Failed to find {target_revision_id}.')
+
+    def upgrade(self, target_revision_id=None):
+        if not self.check_upgrade_id(target_revision_id):
+            logger.info('Data-all version is up to date')
+            return self.current_migration.revision_id()
+
+        logger.info(f"Upgrade from {self.current_migration.revision_id()} to {target_revision_id or 'latest'}")
+        executed_upgrades: Deque[Union[Type[MigrationBase], None]] = deque()
+        saved_start_migration = self.current_migration
+        while self.current_migration is not None:
+            try:
+                logger.info(f'Applying migration {self.current_migration.__name__}')
+                self.current_migration.up()
+                executed_upgrades.append(self.current_migration)
```

Comment:
If we throw on `up` that means that the current_migration won't be added to executed_upgrades. As a result in the `Exception` block `down` will not be executed for it and might leave the environment in limbo. I am not sure what's the correct behaviourm, should we call `down` on the migration that caused the exception on `up`?

Reply:
I think, if it's not executed, then we don't need to downgrade.
If it fails and leaves garbage behind -- I think, it's a bad migration.
If we execute here downgrade, it may as well be unsafe, if poorly written

Reply:
@dlpzx what do you think?

Reply:
If I understand correctly, we are trying to minimize the danger of garbage or halfway done upgrades right? We can either 1) make the migrations atomic - if anything fails in the upgrade it should revert back the changes. Or we can 2) make the migrations retry-able - if anything fails we execute downgrade.Although in theory I think 1) makes more sense, in practice I think it is easier to implement 2), plus it has other advantages (e.g. retry for any other reason). At this moment, the remove_wildcard script fulfills 2 but not 1